### DISCUSSION TITLE
simple collaboration model (to be developed)

### DISCUSSION THREAD

**Discussion by nnako** [freeplane/freeplane#426] [Ideas]
Hi,

as an answer to what we discussed within our last [Freeplane community online conference](https://github.com/freeplane/freeplane/discussions/339), I would like to introduce the idea for a "simple" collaboration mechanism. The realization of this idea will in no way fulfill all the possible collaboration requirements out there, but could be a starting point for further development / refinement.

The first development iteration is to provide the following use cases:

- **edit contents within a map simultaneously** - multiple users might collaborate on a specific mindmap. the activities of all connected collaboration partners will be visible within each local Freeplane editor instance within a red-ish cloud. locally, each user might be able to create, read, update and delete nodes, freely, if they reside within a green-ish cloud.

- **manage reservations** - when multiple users collaborate on a single map, there is the necessity for some kind of access management. thus, users can request / free "reservations" for specific trees within the map. after grant from every collaboration partner, this reserved tree turns green-ish on the requester's local Freeplane editor instance.

Based on the [system structure](https://github.com/freeplane/freeplane/discussions/426#discussioncomment-2885346) (coming within the next post), I'd like to ask for your suggestions, ideas, questions, ... in order to find a realization which satifies as many people as possible.

You will find further information about [the meaning of used symbols](https://github.com/freeplane/freeplane/discussions/426#discussioncomment-2885354) and the [realization concepts](https://github.com/freeplane/freeplane/discussions/426#discussioncomment-2885386) further below within the comments.

The essence of all your comments will be updated up here, so that nobody would be obliged to read all the single comments in order to understand the current state of discussion and development.

**Comment by nnako** [comment]
![220629_freeplane__feat__collaboration](https://user-images.githubusercontent.com/5845602/176369261-bd7ed2bb-466e-4f1e-b213-c3d6d1864a54.png)

**Reply by nnako** [reply]
here come some information about the legend of the concept map:

connections
```
blue lines   - configuration information flow

green lines  - user data flow
```
functionality
```
gray blocks  - activities to be performed within any local Freeplane
               instance at the given time. these activities might be
               realized within the Freeplane code base by Java
               or groovy means.
```
user editor interface
```
red clouds   - indication that here only local read access is granted
               when in active collaboration mode. another user has
               reserved the region / tree.

green clouds - indication that here local read and write operations
               will be valid
```

**Reply by nnako** [reply]
and some words about the concept map

- a given synchronization mechanism is used to synchronize all relevant user and configuration data between different Freeplane editor instances for users participating in a particular collaboration action. here, NextCloud is mentioned, but this could be any other mechanism like Dropbox, Seafile, Syncthing, ... those mechanisms have all reached a good stability.

- all configuration information necessary for handshaking / negotiating registration requests by any collaboration user is stored within simple text files (some user specific and some general) which are shared automatically between all participating Freeplane editor instance.

- each local collaboration user will see colored clouds (red or green) within his own Freeplane editor instance. GREEN indicates that he / she is allowed to create, read, update and delete any nodes within that particular region / tree. RED indicates that there is another collaboration user currently occupying that particular region / tree. Here, only local read operations will be possible.

- any collaboration user may reserve currently un-reserved / free spaces within the collaboration mindmap or request the reservation for currently reserved ones. The collaboration mechanism will figure out which user has to receive a reminder in the form "there is another user waiting for you to get finished editing the tree ... . please free your current reservation, soon". In case of a currently free region, the collaboration mechanism will announce the particular reservation request to all the collaborating entities. They will all display a red cloud for it, locally. The requester turns green for the registered region.

- all edits made locally will not be visible to any other user until they are saved, locally. When saved, the collaboration functionality will check for contents to be synchronized and do it.

**Comment by nnako** [comment]
here is another requirement to be met, when trying to provide collaboration.

**person-specific local translation of file paths**

when working on mindmaps with multiple persons, there is great potential for file paths within link attributes of nodes not reaching their local destination, when clicked. as every collaboration participant might place the mindmap file anywhere within his local system (this is relevant for relative paths) or use entire individual file structures (this is relevant for absolute paths). to solve this, there must be a "file path translation" mechanism which handles these links dynamically and on an individual level, based on each participant's individual file structure settings.

[[see updated concept map](https://github.com/freeplane/freeplane/discussions/426#discussioncomment-2885346)]

**Reply by EdoFro** [reply]
I have a script for that. I can post it here in a few hours. (it’s 7:00 am here)

**Reply by EdoFro** [reply]
here it is. the whole explanation is in the map itself.
[Check in script.zip](https://github.com/freeplane/freeplane/files/8959379/Check.in.script.zip)


hope it helps.

bye, edo

here is the script:

```groovy
dbg = false
if (dbg) println 'Check in v2 - debug on'

def lastLocSto = 'currentFileLocation' //name of the Map Storage Variable that contains last location
def m = node.map //getting the active map
def base = node.pathToRoot.find{ it.style.name == 'baseFolder' } //looks if there is a node that represents the "base folder" of all files in that branch
if (dbg) println "base: ${base}"

def justThisBranch = base ? true : false
if (dbg) println "justThisBranch: ${justThisBranch}"

// return justThisBranch
// looks which was the last folder where the map was when running this script (check in in other computer)
def lastLoc = justThisBranch ? base.link.file.canonicalPath : m.storage[lastLocSto]?.string
if (dbg) println "lastLoc: ${lastLoc}"
// return lastLoc

// looks for the folder of the current map
def currLoc = getCurrentLocation( justThisBranch ? base : node.map )
if (dbg) println "currLoc: ${currLoc}"
// return currLoc

def getCurrentLocation( org.freeplane.plugin.script.proxy.MapProxy mapa ){
    if (dbg) println "getCurrentLocation: mapa: ${mapa}"
    if (dbg) println "  -  mapa.file.parent: ${mapa.file.parent}"
    return mapa.file.parent
}

def getCurrentLocation( org.freeplane.plugin.script.proxy.NodeProxy nodo ){
    if (dbg) println "getCurrentLocation: nodo: ${nodo}"
    def nodoAlternativas = nodo?.children.find{ n -> n.style.name == 'locked' && n.text.toLowerCase() =='possible paths' }
    if (dbg) println "  -  nodoAlternativas: ${nodoAlternativas}"
    def currentLocationNode = nodoAlternativas?.children.find{ n -> n.link.file != null && n.link.file.exists() }
    if (dbg) println "  -  currentLocationNode: ${currentLocationNode}"
    nodoAlternativas?.children*.icons*.clear()
    currentLocationNode?.icons.add('emoji-27A1')
    return currentLocationNode ? currentLocationNode.link.file.canonicalPath : getCurrentLocation( nodo.map )
}

// if the map has no map variable yet, then it gives you the possibility to add it and ends the script
if (!lastLoc) {
    if( ui.showConfirmDialog( node.delegate, "No '$lastLocSto' storage variable in this map.\n\n Do you want to add it to the current map?".toString(), 'Location Map Storage Variable', 0 ) == 0 ){
        m.storage[lastLocSto] = currLoc
        return 'new lastLocSto. ended'
    } else {
        return 'interrupted'
    }
}

// if they are the same then do nothing, there was no travel --> do nothing
if( lastLoc == currLoc ){
    ui.showMessage("current directory is already set to:\n\n $currLoc".toString(),2)
    return 'interrupted'
}

def baseNode = justThisBranch ? base : node.map.root
if (dbg) println "baseNode: ${baseNode}"

// list of all nodes that have a link to a file and are not locked
def nodos = baseNode.find{ n -> n.link.file != null && !n.pathToRoot.any{ it.style.name == 'locked' } }
if (dbg) println "nodos.size(): ${nodos.size()}"
// return nodos.size()

// repeat for ExternalObjects Images
// list of all nodes that have a link to a file
def nodos2 = baseNode.find{ n -> n.externalObject.uri != null && !n.pathToRoot.any{ it.style.name == 'locked' } }
if (dbg) println "nodos2.size(): ${nodos2.size()}"
// return nodos2.size()
// return '\n' + lastLoc + '\n' + currLoc

// change the link for each one of them replacing the last location with the new one
nodos.each{ n ->
    n.link.file = new File( n.link.file.canonicalPath.replace( lastLoc, currLoc ) )
}

// repeat for ExternalObjects Images
// change the link for each one of them replacing the last location with the new one
nodos2.each{ n ->
    def fileX = new File( n.externalObject.uri.toURI() )
    n.externalObject.file = new File( fileX.canonicalPath.replace( lastLoc, currLoc ) )
}

// update the last location information, for the next Check In
if( !justThisBranch ){
    m.storage[lastLocSto] = currLoc
    if (dbg) println "last location information in map storage variable updated: ${currLoc}"
}

// success
ui.informationMessage( "directory changed to:\n\n $currLoc \n\nChange done in ${justThisBranch?'this Basenode branch':'the whole map'}".toString())
c.statusInfo =  "done, changed ${nodos.size()} links + ${nodos2.size()} embeded images  =  ${nodos.size() + nodos2.size()} total"

```

**Reply by EdoFro** [reply]
Sorry, I read the whole Topic again.
I thought you were asking for a script to share a mindmap to organize this project between the people who will collaborate in it. But you were asking for a solution for the "sharing tool" itself.

Maybe it can help as a basis idea.

**Reply by nnako** [reply]
Hi @EdoFro , no problem, thanks for your contribution. We will see how and where either the implementation or the concepts of your code will fit in.

**Comment by EdoFro** [comment]
Dear @nnako,
I think this project idea is super interesting. I also think that is a good idea to make (first) a simple solution that maybe isn't in real time and has some restrictions but that it can be implemented "soon". It is surely better than a thinking in a great implementation that no one will ever do.

Some thoughts:

 - I think it would be convenient to know if you are thinking that this tool should be implemented as a Freeplane modification, an addon or a plugin.
 - I think it would be useful to create a github repository for this project
 There we can also 
    - add some example scripts
    - make some proofs of concept
    - have a dedicated discussion space
    - test different implementation strategies
    - etc.
- Maybe, by knowing who is willing to help, will be easier to choose which solution are we able to code
- I think you have to get a team as soon in the project as possible. It's harder for anyone to jump into a project when it has already started, the decissions have been made and there is little documentation available.

jm2c

edo

**Reply by nnako** [reply]
Hi @EdoFro ,

great thoughts, thanks. just briefly:

- yes, I'd like to start with a very basic implementation which facilitates basic collaboration
- further features should be implemented after continual collection and prioritization, I think
- we already concluded that a plugin might be the best choice (compared to direct modification or plugin)
- yes, a github repo would be helpful for doing the "real work"
- in order to avoid having to read all the old discussions when trying to get on track later, I decided to regularly summarize all technical concepts within one big concept overview as shown in the [1st comment](https://github.com/freeplane/freeplane/discussions/426#discussioncomment-2885346). with this concept map, discussions, collection and prioritization of new featues, ... will become much easier than without it.

the announcement thread, here, is meant to be an introduction of the idea, giving the community a chance to get interested and join in whatever way possible (implementer, tester, documenter, designer,...).

**Comment by nnako** [comment]
here is another requirement to be met, when trying to provide collaboration.

**version-agnostic rendering of XML**

due to different Freeplane versions used by collaboration participants, there is to be a mechanism which translates each relevant part of mindmap structure into a local meta-structure which is then rendered locally. otherwise, trying to merge e.g. a part of a map v1.3.15 and a part of a map v1.9.10 would potentially result in a rendering conflict. due to missing downward compatibility.

[[ see updated concept map ]](https://github.com/freeplane/freeplane/discussions/426#discussioncomment-2885346)

**Comment by f3rrix** [comment]
I found this thread after seeing the repo "freeplane-server" and being sad that it looks like it was started and abandoned years ago. 

I would do *a lot* to support collaborative freeplane (anyone need funding?).  I keep looking at online mindmapping services and so far none of them have any good UIs for serious large-scale map usage, they are all toys.
